functions/api/helpers.js

```diff
@@ -100,3 +107,93 @@ export async function refreshKPILatestValue(kpiRef) {
     });
   }
 }
+
+/**
+ * Build and return KPI response object.
+ *
+ * `kpiSnapshot` is the Firestore KPI document snapshot.
+ */
+export async function buildKpiResponse(kpiSnapshot) {
+  const {
+    created,
+    createdBy,
+    currentValue,
+    edited,
+    editedBy,
+    name,
+    type,
+    updateFrequency,
+  } = kpiSnapshot.data();
+
+  const latestMeasurement = await kpiSnapshot.ref
+    .collection('progress')
+    .orderBy('timestamp', 'desc')
+    .limit(1)
+    .get()
+    .then((snapshot) => {
+      if (!snapshot.docs[0]) return null;
+      const { value, timestamp } = snapshot.docs[0].data();
+      return { value, timestamp: timestamp.toDate() };
+    });
+
+  const isStale = determineKPIStaleState(updateFrequency, latestMeasurement);
+
+  return {
+    currentValue,
+    name,
+    type,
+    lastUpdated: latestMeasurement || null,
+    updateFrequency: updateFrequency || null,
+    isStale,
+    created: created ? created.toDate() : null,
+    createdBy: createdBy ? await getUserDisplayName(createdBy) : null,
+    edited: edited ? edited.toDate() : null,
+    editedBy: editedBy ? await getUserDisplayName(editedBy) : null,
+  };
+}
+
+/**
+ * Return true if a KPI is considered "stale", i.e. no progress measurement
+ * registered within declared update frequency.
+ *
+ * `updateFrequency` is the expected update interval for the KPI.
+ * `progressRecord` is the progress record to check against.
+ */
+function determineKPIStaleState(updateFrequency, progressRecord) {
+  if (!updateFrequency) return null;
+  if (!progressRecord) return true;
+  if (updateFrequency === 'irregular') return false;
+
+  const duration = (frequency) => {
+    switch (frequency) {
+      case 'daily':
+        return { days: 1 };
+      case 'weekly':
+        return { weeks: 1 };
+      case 'monthly':
+        return { months: 1 };
+      case 'quarterly':
+        return { months: 3 };
+      case 'annual':
+        return { years: 1 };
+      default:
+        throw new Error('Unsupported frequency option');
+    }
+  };
+
+  try {
+    const now = new Date();
+    const intervalEnd = set(progressRecord.timestamp, {
```

Comment:
Tenker jeg feil, eller må denne skrus til 23:59:59 for å få med målinger som har skjedd samme dag?

Reply:
Innrømmer at jeg har surret litt rundt med dette, og muligens landet feil: Man går 1 dag, 1 uke, osv. bakover fra _dagens dato_, med _samme tidspunkt siste måling_,  for å sjekke om det har kommet ny måling innenfor det intervallet. Sikrer man ikke da at det har kommet ny siden sist, uavhengig av når målingen er gjort? Hvis man tar utgangspunkt i `endOfDay(now)`, vil ikke intervallet være lenger enn 1 døgn, f.eks?

Reply:
Ah, jeg skjønte ikke at `set` beholder tidspunktet fra `new Date()` (som er nå). Da tror jeg at det blir riktig, ja.

Reply:
Eller blir det mest riktig sånn det er, og heller, når vi skal sjekke og gjøre våre målinger, gjøre det ved slutten av dagen?

Reply:
> Mulig det ikke blir helt riktig å ha et såpass strengt 24-timers intervall (f.eks)? I vår kontekst skal kanskje denne være `false` kl. 16:01 den 18., men først `true` ved midnatt (selv om det er over 24 timer siden siste måling)? thinking
> 
> > mesaurement: 2022-12-17T12:00:00.000Z
> > start: 2022-11-17T12:00:00.000Z
> > end: 2022-11-18T12:00:00.000Z
> > true

Ah, ja, tenker at den burde vente til midnatt. Siden vi kaller det "daglig" og ikke "hver 24. time" burde det være mulig å legge inn målingen når som helst den neste dagen uten at den går stale. Dvs. at man burde kunne ha nesten 48 timers mellomrom mellom to målinger i det mest ekstreme tilfellet.

Reply:
> Ah, ja, tenker at den burde vente til midnatt. Siden vi kaller det "daglig" og ikke "hver 24. time" burde det være mulig å legge inn målingen når som helst den neste dagen uten at den går stale. Dvs. at man burde kunne ha nesten 48 timers mellomrom mellom to målinger i det mest ekstreme tilfellet.

Ja, det blir kanskje det. Skal vi satse på at f.eks.: 
```
>  > mesaurement: 2022-11-20T12:00:00.000Z
>  -----------
>  > now: 2022-11-21T14:07:34.540Z
>  > start: 2022-11-20T00:00:00.000Z
>  > end: 2022-11-21T23:59:59.999Z
>  -----------
>  > stale? false
```
dvs. at intervallet `progressRecord.timestamp` må være innenfor er:
```
{
  start: sub(startOfDay(now), duration(updateFrequency)),
  end: endOfDay(now),
}
```